src/server/initServer.ts

```diff
@@ -267,13 +263,144 @@ const initServer = (
           );
         }
 
+        // Build MongoDB query based on filters
+        const query: any = {};
+
+        /* -------------- Date Filter ------------- */
+        const { startDate, endDate } = filters.dateFilterState;
+        const startTimestamp = new Date(startDate.year, startDate.month - 1, startDate.day).getTime();
+        const endTimestamp = new Date(endDate.year, endDate.month - 1, endDate.day + 1).getTime() - 1;
+
+        query.timestamp = {
+          $gte: startTimestamp,
+          $lte: endTimestamp,
+        };
+
+        /* ------------ Context Filter ------------ */
+        const { contextFilterState } = filters;
+        const contextConditions: any[] = [];
+
+        Object.keys(contextFilterState).forEach((context) => {
+          const value = contextFilterState[context];
+          if (typeof value === 'boolean') {
+            if (value) {
+              // The entire context is selected
+              contextConditions.push({ context });
+            }
+          } else {
+            // The context has subcontexts
+            const subcontexts = Object.keys(value).filter((subcontext) => { return value[subcontext]; });
+            if (subcontexts.length > 0) {
+              contextConditions.push({
+                context,
+                subcontext: { $in: subcontexts },
+              });
+            }
+          }
+        });
+
+        if (contextConditions.length > 0) {
+          query.$or = contextConditions;
+        }
+
+        /* -------------- Tag Filter -------------- */
+        const { tagFilterState } = filters;
+        const selectedTags = Object.keys(tagFilterState).filter((tag) => { return tagFilterState[tag]; });
+        if (selectedTags.length > 0) {
+          query.tags = { $in: selectedTags };
+        }
+
+        /* --------- Action/Error Filter ---------- */
+        const { actionErrorFilterState } = filters;
+
+        if (actionErrorFilterState.type) {
+          query.type = actionErrorFilterState.type;
+        }
+
+        if (actionErrorFilterState.type === 'error' && actionErrorFilterState.errorMessage) {
```

Comment:
Let's move the filter types to the server, then `npm run copy-server-types` to get them on the client. Then, reference the types: `actionErrorFilterState.type === ActionType.Error`